services/httpd/handler.go

```diff
@@ -760,6 +777,277 @@ func convertToEpoch(r *query.Result, epoch string) {
+		bs = make([]byte, 0, r.ContentLength)
+	}
+	buf := bytes.NewBuffer(bs)
+
+	_, err := buf.ReadFrom(body)
+	if err != nil {
+		if err == errTruncated {
+			h.httpError(w, http.StatusText(http.StatusRequestEntityTooLarge), http.StatusRequestEntityTooLarge)
+			return
+		}
+
+		if h.Config.WriteTracing {
+			h.Logger.Info("Prom write handler unable to read bytes from request body")
+		}
+		h.httpError(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	atomic.AddInt64(&h.stats.WriteRequestBytesReceived, int64(buf.Len()))
+
+	if h.Config.WriteTracing {
+		h.Logger.Info(fmt.Sprintf("Prom write body received by handler: %s", buf.Bytes()))
+	}
+
+	reqBuf, err := snappy.Decode(nil, buf.Bytes())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// Convert the Prometheus remote write request to Influx Points
+	var req remote.WriteRequest
+	if err := proto.Unmarshal(reqBuf, &req); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	points, err := prometheus.WriteRequestToPoints(&req)
+	if err != nil {
+		h.Logger.Info(fmt.Sprintf("promerr: %s", err.Error()))
+		if h.Config.WriteTracing {
+			h.Logger.Info(fmt.Sprintf("Prom write handler: %s", err.Error()))
+		}
+
+		if err != prometheus.ErrNaNDropped {
+			h.httpError(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+	}
+
+	// Determine required consistency level.
+	level := r.URL.Query().Get("consistency")
+	consistency := models.ConsistencyLevelOne
+	if level != "" {
+		var err error
+		consistency, err = models.ParseConsistencyLevel(level)
+		if err != nil {
+			h.httpError(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+	}
+
+	// Write points.
+	if err := h.PointsWriter.WritePoints(database, r.URL.Query().Get("rp"), consistency, user, points); influxdb.IsClientError(err) {
+		atomic.AddInt64(&h.stats.PointsWrittenFail, int64(len(points)))
+		h.httpError(w, err.Error(), http.StatusBadRequest)
+		return
+	} else if influxdb.IsAuthorizationError(err) {
+		atomic.AddInt64(&h.stats.PointsWrittenFail, int64(len(points)))
+		h.httpError(w, err.Error(), http.StatusForbidden)
+		return
+	} else if werr, ok := err.(tsdb.PartialWriteError); ok {
+		atomic.AddInt64(&h.stats.PointsWrittenOK, int64(len(points)-werr.Dropped))
+		atomic.AddInt64(&h.stats.PointsWrittenDropped, int64(werr.Dropped))
+		h.httpError(w, werr.Error(), http.StatusBadRequest)
+		return
+	} else if err != nil {
+		atomic.AddInt64(&h.stats.PointsWrittenFail, int64(len(points)))
+		h.httpError(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	atomic.AddInt64(&h.stats.PointsWrittenOK, int64(len(points)))
+	h.writeHeader(w, http.StatusNoContent)
+}
+
+// servePromRead will convert a Prometheus remote read request into an InfluxQL query and
+// return data in Prometheus remote read protobuf format.
+func (h *Handler) servePromRead(w http.ResponseWriter, r *http.Request, user meta.User) {
+	compressed, err := ioutil.ReadAll(r.Body)
```

Comment:
Every `http.Error` in here should be a `h.httpError`.

Reply:
Addressed in f0ffe201615dc9d75c9a6fd064faa076eb99eb16

Reply:
Is it appropriate to use `h.httpError` here? That will output the error in a format for InfluxDB. While at the moment we use JSON, it could be any format like CSV. If we are attempting to send an error back that matches what Prometheus is expecting, it probably isn't correct to use `h.httpError`.

Reply:
The Prometheus remote client doesn't do anything with the response body on error so it's probably not an issue: https://github.com/prometheus/prometheus/blob/master/storage/remote/client.go#L123-L134